xmppserver/src/main/java/org/jivesoftware/openfire/net/StanzaHandler.java

```diff
@@ -484,7 +490,25 @@ protected boolean negotiateTLS() {
      * depending on the session type such as auth for Non-SASL authentication and register
      * for in-band registration.
      */
-    protected void tlsNegotiated() {
+    protected void tlsNegotiated(XmlPullParser xpp) throws XmlPullParserException, IOException {
+        // Discard domain values obtained from unencrypted initiating stream
+        // in accordance with RFC 6120 § 5.4.3.3. See https://datatracker.ietf.org/doc/html/rfc6120#section-5.4.3.3
+        this.domainPair = null;
+
+        // Get remoteDomain from encrypted stream tag
+        for (int eventType = xpp.getEventType(); eventType != XmlPullParser.START_TAG;) {
+            eventType = xpp.next();
+        }
+        this.domainPair = new DomainPair(XMPPServer.getInstance().getServerInfo().getXMPPDomain(), xpp.getAttributeValue("", "from"));
+
+        // Discard session data obtained from unencrypted initiating stream
+        // in accordance with RFC 6120 § 5.4.3.3. See https://datatracker.ietf.org/doc/html/rfc6120#section-5.4.3.3
+        SessionManager sessionManager = SessionManager.getInstance();
+        sessionManager.unregisterIncomingServerSession(session.getStreamID());
```

Comment:
This is server-to-server specific. It should not be defined in an abstract class, as that shares this functionality with not onlty the Server-to-Server subclass, but also the Client-to-Server, ExternalComponent and Multiplexer subclasses.

Reply:
Moved to ServerStanzaHandler